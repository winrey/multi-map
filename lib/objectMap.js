"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectMultiMap = void 0;
const arrayMap_1 = require("./arrayMap");
/**
 * 注意，这个没有顺序区分
 */
class ObjectMultiMap {
    constructor(keyIndex, opts = {
        ignoredExtarField: false,
    }) {
        this._keyIndex = this.geneKeysIndex(keyIndex);
        this._data = new arrayMap_1.ArrayMultiMap(this._keyIndex.length);
        this._opts = opts;
    }
    get [Symbol.toStringTag]() {
        return 'ObjectMultiMap';
    }
    *[Symbol.iterator]() {
        for (const result of this.entries()) {
            yield result;
        }
    }
    get size() {
        return this._data.size;
    }
    get keyIndex() {
        return this._keyIndex;
    }
    geneKeysIndex(key) {
        if (!key || !['object', 'array'].includes(typeof key)) {
            throw new TypeError('keyIndex type invalid');
        }
        if (!(key instanceof Array)) {
            key = Object.keys(key);
        }
        if (!key.length) {
            throw new TypeError('keyIndex lens invalid');
        }
        return key;
    }
    checkAndParseKeys(key, ignoredExtarField) {
        /* istanbul ignore next */
        ignoredExtarField = ignoredExtarField !== null && ignoredExtarField !== void 0 ? ignoredExtarField : this._opts.ignoredExtarField;
        if (typeof key !== 'object' || key instanceof Array) {
            throw new TypeError('key should be an obj');
        }
        let kkey = Object.keys(key);
        this._keyIndex.forEach((v) => {
            const filtered = kkey.filter((k) => k != v);
            if (kkey.length === filtered.length) {
                throw new TypeError(`index field "${v}" not found`);
            }
            kkey = filtered;
        });
        if (!ignoredExtarField && kkey.length) {
            throw new TypeError(`key have extra fields: ${JSON.stringify(kkey)}`);
        }
        return this._keyIndex.map((k) => key[k]);
    }
    index2Keys(index) {
        const result = {};
        /* istanbul ignore if  */
        if (index.length !== this._keyIndex.length) {
            throw new Error('index lens not matched');
        }
        for (let i = 0; i < index.length; i++) {
            result[this._keyIndex[i]] = index[i];
        }
        return result;
    }
    set(keys, data) {
        const k = this.checkAndParseKeys(keys);
        this._data.set(k, data);
        return this;
    }
    get(keys, defaultVal) {
        const k = this.checkAndParseKeys(keys);
        return this._data.get(k, defaultVal);
    }
    has(keys) {
        const k = this.checkAndParseKeys(keys);
        return this._data.has(k);
    }
    delete(keys) {
        const k = this.checkAndParseKeys(keys);
        return this._data.delete(k);
    }
    clone() {
        const clone = new ObjectMultiMap(this._keyIndex);
        for (const [k, v] of this.entries()) {
            clone.set(k, v);
        }
        return clone;
    }
    clear() {
        this._data.clear();
    }
    forEach(callbackfn, thisArg) {
        this._data.forEach((v, k) => {
            const keys = this.index2Keys(k);
            callbackfn.apply(thisArg, [v, keys, this]);
        });
    }
    *entries() {
        for (const [k, v] of this._data.entries()) {
            const keys = this.index2Keys(k);
            yield [keys, v];
        }
    }
    *keys() {
        for (const result of this.entries()) {
            yield result[0];
        }
    }
    *values() {
        for (const result of this.entries()) {
            yield result[1];
        }
    }
}
exports.ObjectMultiMap = ObjectMultiMap;
