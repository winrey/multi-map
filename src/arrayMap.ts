/**
 * 注意，这个没有顺序区分
 */
export class ArrayMultiMap<TKeys extends any[] = any[], TValue = string> implements Map<TKeys, TValue> {
  private _size = 0;
  private _data = new Map<any, any>();
  private _leafValue = Symbol('leafValue');
  private _defaultFunc?: (keys: TKeys) => TValue;

  constructor(defaultFunc?: (keys: TKeys) => TValue) {
    this._defaultFunc = defaultFunc;
  }

  get [Symbol.toStringTag]() {
    return 'ArrayMultiMap';
  }

  *[Symbol.iterator](): IterableIterator<[TKeys, TValue]> {
    for (const result of this.entries()) {
      yield result;
    }
  }

  get size() {
    return this._size;
  }

  clone() {
    const clone = new ArrayMultiMap<TKeys, TValue>();
    for (const [k, v] of this.entries()) {
      clone.set(k, v);
    }
    return clone;
  }

  set(keys: TKeys, data: TValue) {
    const map = keys.reduce((last: Map<any, any>, key, i) => {
      if (last.has(key)) {
        return last.get(key) as Map<any, any>;
      }
      const map = new Map<any, any>();
      last.set(key, map);
      return map;
    }, this._data);
    if (!map.has(this._leafValue)) {
      this._size += 1;
    }
    map.set(this._leafValue, data);
    return this;
  }

  get(keys: TKeys, defaultVal?: TValue): TValue | undefined {
    const exit = Symbol('not found');
    const val = keys.reduce((last: Map<any, any> | symbol, key, i) => {
      if (last === exit) {
        return exit;
      }
      last = last as Map<any, any>;
      if (!last.has(key)) {
        return exit;
      }
      return last.get(key) as Map<any, any>;
    }, this._data);
    if (val === exit || !val.has(this._leafValue)) {
      const result = defaultVal ?? this._defaultFunc?.(keys);
      if (result !== undefined) {
        this.set(keys, result);
      }
      return result;
    }
    return val.get(this._leafValue);
  }

  has(keys: TKeys): boolean {
    const exit = Symbol('not found');
    const lastMap = keys.reduce((last: Map<any, any> | symbol, key, i) => {
      if (last === exit) {
        return exit;
      }
      last = last as Map<any, any>;
      if (!last.has(key)) {
        return exit;
      }
      return last.get(key) as Map<any, any>;
    }, this._data);
    if (lastMap === exit) {
      return false;
    }
    return lastMap.has(this._leafValue);
  }

  delete(keys: TKeys) {
    const exit = Symbol('not found');
    const routeStack: Map<any, any>[] = [this._data];
    const result = keys.reduce((last: Map<any, any> | symbol, key, i) => {
      if (last === exit) {
        return exit;
      }
      last = last as Map<any, any>;
      if (!last.has(key)) {
        return exit;
      }
      const next = last.get(key) as Map<any, any>;
      routeStack.push(next);
      return next;
    }, this._data);

    if (result === exit || !result.has(this._leafValue)) {
      return false;
    }

    result.delete(this._leafValue);
    this._size -= 1;

    // clear unnecessary map
    for (let i = routeStack.length - 1; i > 0; i--) {
      if (routeStack[i].size) {
        break;
      }
      // need clear
      routeStack[i - 1].delete(keys[i - 1]);
    }
    return true;
  }

  clear(): void {
    this._size = 0;
    this._data = new Map<any, any>();
  }

  forEach(callbackfn: (value: TValue, keys: TKeys, map: Map<TKeys, TValue>) => void, thisArg?: any) {
    const foreachIfMap = (map: Map<any, any>, keys: any[] = []) => {
      map.forEach((v, k) => {
        if (k === this._leafValue) {
          callbackfn.apply(thisArg ?? this, [v, [...keys] as TKeys, this]);
          return;
        }
        foreachIfMap(v, [...keys, k]);
      });
    };
    foreachIfMap(this._data);
  }

  *entries(): IterableIterator<[TKeys, TValue]> {
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    const that = this;
    const foreachIfMap = function* (map: Map<any, any>, keys: any[] = []): any {
      for (const [k, v] of map) {
        if (k === that._leafValue) {
          yield [[...keys], v];
          break;
        }
        for (const result of foreachIfMap(v, [...keys, k])) {
          yield result;
        }
      }
    };
    for (const result of foreachIfMap(this._data)) {
      yield result;
    }
  }

  *keys(): IterableIterator<TKeys> {
    for (const result of this.entries()) {
      yield result[0];
    }
  }

  *values(): IterableIterator<TValue> {
    for (const result of this.entries()) {
      yield result[1];
    }
  }
}
