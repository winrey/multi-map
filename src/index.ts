export * from './arrayMap';
export * from './objectMap';
