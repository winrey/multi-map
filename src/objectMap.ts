import { ArrayMultiMap } from "./arrayMap";

export type TOptions = {
  ignoredExtarField?: boolean
}

/**
 * 注意，这个没有顺序区分
 */
export class ObjectMultiMap <TKey extends Record<string, any> = Record<string, any>, TValue=string> implements Map<TKey, TValue> {

  private _keyIndex: string[]
  private _data: ArrayMultiMap<any[], TValue>
  private _opts: TOptions

  get [Symbol.toStringTag]() {
    return "ObjectMultiMap";
  }

  *[Symbol.iterator](): IterableIterator<[TKey, TValue]> {
    for (const result of this.entries()) {
      yield result
    }
  }

  get size () {
    return this._data.size
  }

  get keyIndex () {
    return this._keyIndex
  }

  constructor(keyIndex: Array<string> | Record<string, any>, opts: TOptions = {
    ignoredExtarField: false,
  }) {
    this._keyIndex = this.geneKeysIndex(keyIndex)
    this._data = new ArrayMultiMap<any[], TValue>(this._keyIndex.length)
    this._opts = opts
  }

  private geneKeysIndex(key?: Array<string> | Record<string, any>) {
    if (!key || !["object", "array"].includes(typeof key)) {
      throw new TypeError("keyIndex type invalid")
    }
    if (!(key instanceof Array)) {
      key = Object.keys(key)
    }
    if (!key.length) {
      throw new TypeError("keyIndex lens invalid")
    }
    return key as string[]
  }

  private checkAndParseKeys(key: TKey, ignoredExtarField?: boolean) {
    ignoredExtarField = ignoredExtarField ?? this._opts.ignoredExtarField
    if (typeof key !== "object" || key instanceof Array) {
      throw new TypeError("key should be an obj")
    }
    let kkey = Object.keys(key)
    this._keyIndex.forEach(v => {
      const filtered = kkey.filter(k => k != v)
      if (kkey.length === filtered.length) {
        throw new TypeError(`index field "${v}" not found`)
      }
      kkey = filtered
    })
    if (!ignoredExtarField && kkey.length) {
      throw new TypeError(`key have extra fields: ${JSON.stringify(kkey)}`);
    }
    return this._keyIndex.map(k => key[k]) 
  }

  private index2Keys(index: any[]) {
    const result: Record<string, any> = {}
    /* istanbul ignore if  */
    if (index.length !== this._keyIndex.length) {
      throw new Error("index lens not matched")
    }
    for (let i = 0; i < index.length; i++) {
      result[this._keyIndex[i]] = index[i]
    }
    return result as TKey
  }

  set(keys: TKey, data: TValue) {
    const k = this.checkAndParseKeys(keys)
    this._data.set(k, data)
    return this
  }

  get(keys: TKey, defaultVal?: TValue): TValue {
    const k = this.checkAndParseKeys(keys)
    return this._data.get(k, defaultVal)
  }

  has(keys: TKey): boolean {
    const k = this.checkAndParseKeys(keys)
    return this._data.has(k)
  }

  delete(keys: TKey) {
    const k = this.checkAndParseKeys(keys)
    return this._data.delete(k)
  }

  clear(): void {
    this._data.clear()
  }

  
  forEach(callbackfn: (value: TValue, keys: TKey, map: Map<TKey, TValue>) => void, thisArg?: any) {
    this._data.forEach((v, k) => {
      const keys = this.index2Keys(k)
      callbackfn.apply(thisArg, [v, keys, this])
    })
  }

  *entries(): IterableIterator<[TKey, TValue]> {
    for (const [k, v] of this._data.entries()) {
      const keys = this.index2Keys(k)
      yield [keys, v]
    }
  }

  *keys(): IterableIterator<TKey> {
    for (const result of this.entries()) {
      yield result[0]
    }
  }

  *values(): IterableIterator<TValue> {
    for (const result of this.entries()) {
      yield result[1]
    }
  }
}
