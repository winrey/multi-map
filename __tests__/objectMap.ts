
import { ObjectMultiMap } from '../src/objectMap';

const obj1 = {
  openId: 6,
  appId: 4,
}
const obj2 = {
  openId: 6,
  appId: 5,
}
const obj3 = {
  openId: 4,
  appId: 5,
}

it("could be created", () => {
  new ObjectMultiMap(["233"])
  new ObjectMultiMap({233: "", "666": ""})
})

it("should check input error", () => {
  expect(() => new ObjectMultiMap(-1 as any)).toThrow()
  expect(() => new ObjectMultiMap(0 as any)).toThrow()
  expect(() => new ObjectMultiMap("test" as any)).toThrow()
  expect(() => new ObjectMultiMap(undefined as any)).toThrow()
  expect(() => new ObjectMultiMap([])).toThrow()
  expect(() => new ObjectMultiMap({})).toThrow()
})

it("could be set and get", () => {
  const map = new ObjectMultiMap(["openId", "appId"])
  expect(map.has(obj1)).toBeFalsy()
  map.set(obj1, "233")
  map.set(obj2, "999")
  expect(map.has(obj1)).toBeTruthy()

  expect(map.has(obj3)).toBeFalsy()
  expect(map.has(obj2)).toBeTruthy()
  expect(map.has({
    appId: 4,
    openId: 6,
  })).toBeTruthy()
  expect(map.get(obj1)).toBe("233")
  expect(map.get(obj1, "888")).toBe("233")
  expect(map.get({
    appId: 4,
    openId: 6,
  })).toBe("233")
  expect(map.get(obj2)).toBe("999")
  expect(map.get(obj2, "888")).toBe("999")
  expect(map.get(obj3)).toBe(undefined)
  expect(map.get(obj3, "888")).toBe("888")
})

it("should throw if lens doesn't match", () => {
  const map = new ObjectMultiMap(["openId", "appId"])
  expect(() => map.set({}, "")).toThrow()
  expect(() => map.set({"openId": 234}, "")).toThrow()
  expect(() => map.get({"openId": 234, "unionId": 3444})).toThrow()
  expect(() => map.has({ openId: 234, unionId: 3444})).toThrow()
  expect(() => map.has({ openId: 234, appId: 233, unionId: 3444})).toThrow()
})

it("could delete and clear", () => {
  const map = new ObjectMultiMap(obj1)
  expect(map.size).toBe(0)
  map.set(obj1, "233")
  map.set(obj2, "2333")

  expect(map.size).toBe(2)
  expect(map.get(obj1)).toBe("233")
  expect(map.has(obj1)).toBeTruthy()

  expect(map.delete(obj1)).toBeTruthy()
  expect(map.delete(obj3)).toBeFalsy()
  
  expect(map.size).toBe(1)
  expect(map.has(obj1)).toBeFalsy()
  expect(map.get(obj1)).toBe(undefined)

  expect(map.has(obj2)).toBeTruthy()

  map.clear()
  expect(map.size).toBe(0)

  expect(map.has(obj2)).toBeFalsy()
})

it("could be foreach", () => {
  const map = new ObjectMultiMap(obj1)
  map.set(obj1, "233")
  map.set(obj2, "2333")
  map.set(obj3, "23333")

  map.forEach((v, k) => {
    expect(v).toMatchSnapshot()
    expect(k).toMatchSnapshot()
  })
})

it("could be iterable", () => {
  const map = new ObjectMultiMap(obj1)
  map.set(obj1, "233")
  map.set(obj2, "2333")
  map.set(obj3, "23333")

  for (const result of map) {
    expect(result).toMatchSnapshot()
  }
})

it("entries could be for iterable", () => {
  const map = new ObjectMultiMap(obj1)
  map.set(obj1, "233")
  map.set(obj2, "2333")
  map.set(obj3, "23333")

  expect(map.entries()).toMatchSnapshot()

  for (const result of map.entries()) {
    expect(result).toMatchSnapshot()
  }
})

it("keys could be for iterable", () => {
  const map = new ObjectMultiMap(obj1)
  map.set(obj1, "233")
  map.set(obj2, "2333")
  map.set(obj3, "23333")

  expect(map.keys()).toMatchSnapshot()

  for (const result of map.keys()) {
    expect(result).toMatchSnapshot()
  }
})

it("values could be for iterable", () => {
  const map = new ObjectMultiMap(obj1)
  map.set(obj1, "233")
  map.set(obj2, "2333")
  map.set(obj3, "23333")

  expect(map.values()).toMatchSnapshot()

  for (const result of map.values()) {
    expect(result).toMatchSnapshot()
  }
})


