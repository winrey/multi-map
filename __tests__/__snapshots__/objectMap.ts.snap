// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`could be foreach 1`] = `"233"`;

exports[`could be foreach 2`] = `
Object {
  "appId": 4,
  "openId": 6,
}
`;

exports[`could be foreach 3`] = `"2333"`;

exports[`could be foreach 4`] = `
Object {
  "appId": 5,
  "openId": 6,
}
`;

exports[`could be foreach 5`] = `"23333"`;

exports[`could be foreach 6`] = `
Object {
  "appId": 5,
  "openId": 4,
}
`;

exports[`could be iterable 1`] = `
Array [
  Object {
    "appId": 4,
    "openId": 6,
  },
  "233",
]
`;

exports[`could be iterable 2`] = `
Array [
  Object {
    "appId": 5,
    "openId": 6,
  },
  "2333",
]
`;

exports[`could be iterable 3`] = `
Array [
  Object {
    "appId": 5,
    "openId": 4,
  },
  "23333",
]
`;

exports[`entries could be for iterable 1`] = `Object {}`;

exports[`entries could be for iterable 2`] = `
Array [
  Object {
    "appId": 4,
    "openId": 6,
  },
  "233",
]
`;

exports[`entries could be for iterable 3`] = `
Array [
  Object {
    "appId": 5,
    "openId": 6,
  },
  "2333",
]
`;

exports[`entries could be for iterable 4`] = `
Array [
  Object {
    "appId": 5,
    "openId": 4,
  },
  "23333",
]
`;

exports[`keys could be for iterable 1`] = `Object {}`;

exports[`keys could be for iterable 2`] = `
Object {
  "appId": 4,
  "openId": 6,
}
`;

exports[`keys could be for iterable 3`] = `
Object {
  "appId": 5,
  "openId": 6,
}
`;

exports[`keys could be for iterable 4`] = `
Object {
  "appId": 5,
  "openId": 4,
}
`;

exports[`values could be for iterable 1`] = `Object {}`;

exports[`values could be for iterable 2`] = `"233"`;

exports[`values could be for iterable 3`] = `"2333"`;

exports[`values could be for iterable 4`] = `"23333"`;
